packages/openlmis/src/Utils.js

```diff
@@ -0,0 +1,93 @@
+import { composeNextState } from '@openfn/language-common';
+import {
+  request as commonRequest,
+  logResponse,
+  makeBasicAuthHeader,
+} from '@openfn/language-common/util';
+
+export const authorize = state => {
+  const { baseUrl, access_token, username, password, clientId, clientSecret } =
+    state.configuration;
+
+  const headers = makeBasicAuthHeader(clientId, clientSecret);
+
+  if (access_token) return state;
+
+  if (username && password) {
+    const options = {
+      query: { grant_type: 'password', username, password },
+      headers: {
+        'Content-Type': 'application/json',
+        ...headers,
+      },
+      baseUrl,
+      parseAs: 'json',
+      maxRedirections: 1,
+    };
+    return commonRequest('POST', '/api/oauth/token', options)
+      .then(response => {
+        return {
+          ...state,
+          configuration: {
+            ...state.configuration,
+            access_token: response.body.access_token,
+          },
+        };
+      })
+      .catch(err => {
+        console.error('Error authenticating with OpenLMIS');
+        console.log(err.body);
+        throw err;
+      });
+  }
+
+  // warn if no auth found
+  console.warn(
+    `WARNING: no authentication properties found on configuration schema!`
+  );
+  console.warn(
+    'Make sure to either set an access_token or username, password, clientId, and clientSecret'
+  );
+};
+
+export const prepareNextState = (state, response, callback = s => s) => {
+  const { body, ...responseWithoutBody } = response;
+
+  const nextState = {
+    ...composeNextState(state, response.body),
+    response: responseWithoutBody,
+  };
+
+  return callback(nextState);
+};
+
+export const request = (configuration = {}, method, path, options) => {
+  const { baseUrl, access_token } = configuration;
+  const { headers = {}, ...otherOptions } = options;
+
+  urlMatchesBase(path, baseUrl);
+
+  const opts = {
+    parseAs: 'json',
+    baseUrl: `${baseUrl}/api`,
+    headers: {
+      ...headers,
+      'content-type': 'application/json',
+      Authorization: `Bearer ${access_token}`,
+    },
+    ...otherOptions,
+  };
+
+  return commonRequest(method, path, opts);
+};
+
+function urlMatchesBase(path, baseUrl) {
+  const base = new URL(baseUrl);
+  const url = new URL(path, baseUrl);
```

Comment:
Is this right?

If path is `https://www.google.com` and base url is `http://openlmis.org`, I suspect `new URL` will throw

I'd like to see a bunch of unit tests against this function to be satisfied that it's working correctly

Reply:
It does throw, I have added one unit test. Can you recommend more test scenarios ?

Reply:
I see 🤔, It's a tricky one 

I think the best way to think about it is as long as your authenticated to a service you can't use any absolute URL with exception of adaptor that already export common http. Those one can still make request to other service. Eg `http.get({url:'https://swapi.dev/api/'})`. 

I am saying authenticated because in http adaptor for example you can use the adaptor without a authentication to any service

Reply:
Yeah, agreed.

For any adaptor with a baseURL (including http), don't allow absolute URLs to be fetched at all.

I've worked out a plan this morning and I'm going to go off and action it ASAP. So we'll keep this here for now and if it changes in a few days, no problem